### PR TITLE
AMOV-IMU codes for px4 1.8.0 rc0

### PR DESCRIPTION
**Pull request by zarathustr**
Dear All

We have completed the code for our AMOV-IMU which consists of high-precision jitter-rejection inertial measurement unit using ADI analog sensors. These sensors have been significantly calibrated for the ADC biases, misalignment, scale factors, biases over temperature and etc. The sensor readings are captured at the frequency of 10KHz via the onboard AD7689. They are later processed by the onboard STM32F4 micro controller for low-pass filtering and application of calibration parameters. We offer two types of connection i.e. SPI and UART. For SPI, only Pixhawk and Pixhawk 4 support such port. While for the UART, most of the autopilots are capable. We especially added the startup code for AMOV-IMU in the rc.sensors hoping that this can be automatically enabled by the SYS_COMPANION parameter. As now the data transfers the IMU data at 1250Hz on the wire with baudrate of 1500000, I suggest the SYS_COMPANION for this case be 1500000 for easy clarification.

The original codes are based on the px4-1.6.2 version and now we have made it fully support the px4-1.8.0 for the new-coming update. The UART driver mostly receives the data via the normal DMA mode with the support of high-priority threads as all of them are DMA-enabled as described in nuttx/config.h.

The performance of the sensors has been fully evaluated. We have done 24-hour and 7x24 hour continuous test where 3 AMOV-IMU are simutaneously sampled using the Pixhawk or Pixhawk2. During the tests, from starting time to the end, there have been not any sensor reading errors found out manually or automatically by the sensors app. So the final conclusion can be drawn that this sensor is highly reliable and is much more accurate facing the vibration. The price is quite low i.e. 1999 RMB (317.7707 USD) for basic version (ADXRS620 + ADXL203) and 2999 RMB (476.7355 USD) for medium version (2x ADXRS620 1x ADXRS642 + ADXL203). For more details in flight, please check https://github.com/PX4/Firmware/issues/9297 and https://github.com/ArduPilot/ardupilot/issues/8211.

Thanks!

### CODE REVIEW

**Comment by huangwen0907** [comment]
you have the build error when build px4fmu-v2_default. 
I think you can comment some unused code.
> /opt/gcc/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld: nuttx_px4fmu-v2_default.elf section `.data' will not fit in region `flash'
/opt/gcc/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld: region `flash' overflowed by 2000 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
Makefile:148: recipe for target 'nuttx_px4fmu-v2_default' failed
make: *** [nuttx_px4fmu-v2_default] Error 1
script returned exit code 2

**Comment by zarathustr** [comment]
@huangwen0907  This is not a failure. Just because of the improper ld.script as the code is too big to fit in the designed area for STM32F407VG. Turn to @LorenzMeier for some tutorial.

**Comment by dagar** [comment]
@zarathustr I would include it everywhere except px4fmu-v2_default for now. Most newer fmu-v2 boards are actually what we call fmu-v3 that use the full 2MB of flash.

**Comment by zarathustr** [comment]
@dgar is it possible to let the bootloader detect the size of the flash so that the autopilot can dynamically link the library?

**Comment by dagar** [comment]
Yes we already do that, although you need bootloader version 5 or newer. Try to flash with QGC and note the bootloader version and reported flash size. If the full 2MB is accessible then it will automatically download and flash px4fmu-v3 instead of px4fmu-v2.

File: src/drivers/boards/px4fmu-v2/board_config.h [hunk 1]
```diff
@@ -215,6 +215,7 @@
 #define PX4_SPIDEV_EXT_ACCEL_MAG PX4_MK_SPI_SEL(PX4_SPI_BUS_EXT, 2)
 #define PX4_SPIDEV_EXT_BARO      PX4_MK_SPI_SEL(PX4_SPI_BUS_EXT, 3)
 #define PX4_SPIDEV_EXT_MPU       PX4_MK_SPI_SEL(PX4_SPI_BUS_EXT, 4)
```
Comment (dagar) [inline src/drivers/boards/px4fmu-v2/board_config.h]: Using PX4_SPIDEV_EXT_MPU should be fine here. We should probably have a better generic external name.

File: src/drivers/imu/amov_imu/amov_imu.cpp [hunk 2]
```diff
@@ -0,0 +1,1560 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2012-2018 PX4 Development Team. All rights reserved.
```
Comment (dagar) [inline src/drivers/imu/amov_imu/amov_imu.cpp]: Minor, copyrate start date.

File: src/drivers/imu/amov_imu/amov_imu.cpp [hunk 3]
```diff
@@ -0,0 +1,1560 @@
...
+	enum Rotation               _rotation;
+
+
+	// last temperature reading for print_info()
+	float                       _last_temperature;
+
+	// keep last accel reading for duplicate detection
+	uint16_t                    _last_accel[3];
+	bool                        _got_duplicate;
+    int                         _task;
```
Comment (dagar) [inline src/drivers/imu/amov_imu/amov_imu.cpp]: Formatting. Try installing astyle and running `make check_format` locally. Once the CI system gets passed the px4fmu-v2_default build it will still fail on these minor formatting errors.

File: cmake/configs/nuttx_px4fmu-v2_default.cmake [hunk 4]
```diff
@@ -24,14 +24,15 @@ set(config_module_list
...
-	drivers/imu/l3gd20
+        drivers/imu/l3gd20
 	drivers/imu/lsm303d
 	drivers/magnetometer/hmc5883
 	drivers/magnetometer/lis3mdl
 	#drivers/mb12xx
 	#drivers/mkblctrl
-	drivers/imu/mpu6000
-	drivers/imu/mpu9250
+        drivers/imu/mpu6000
```
Comment (dagar) [inline cmake/configs/nuttx_px4fmu-v2_default.cmake]: Can you undo these whitespace changes? PX4 Firmware generally uses tabs everywhere. See the astyle comment below.

File: Tools/px_uploader.py [hunk 5]
```diff
@@ -540,9 +540,9 @@ def upload(self, fw, force=False, boot_delay=None):
                 # and thus can support the full flash, see
                 # https://github.com/PX4/Firmware/blob/master/src/drivers/boards/common/stm32/board_mcu_version.c#L125-L144
 
-                if self.fw_maxsize > fw.property('image_maxsize') and not force:
-                    raise RuntimeError("Board can accept larger flash images (%u bytes) than board config (%u bytes). Please use the correct board configuration to avoid lacking critical functionality."
-                        % (self.fw_maxsize, fw.property('image_maxsize')))
+                #if self.fw_maxsize > fw.property('image_maxsize') and not force:
```
Comment (dagar) [inline Tools/px_uploader.py]: Undo

File: ROMFS/px4fmu_common/init.d/rc.sensors [hunk 6]
```diff
@@ -82,6 +82,13 @@ then
 
 	# Internal SPI bus ICM-20608-G
 	mpu6000 -T 20608 start
+	
+	if param compare SYS_COMPANION 1500000
+	then
+		if amov_imu start -X 3 -T 1 -I 2 -A -P 255
```
Comment (dagar) [inline ROMFS/px4fmu_common/init.d/rc.sensors]: This doesn't need to be wrapped in a conditional.
Do you want to add a comment explaining these options? Could some of them be defaults instead?

File: src/drivers/imu/mpu6000/mpu6000.cpp [hunk 7]
```diff
@@ -738,7 +738,7 @@ MPU6000::init()
 
 	/* measurement will have generated a report, publish */
 	_accel_topic = orb_advertise_multi(ORB_ID(sensor_accel), &arp,
-					   &_accel_orb_class_instance, (is_external()) ? ORB_PRIO_MAX : ORB_PRIO_HIGH);
+                       &_accel_orb_class_instance, (is_external()) ? ORB_PRIO_DEFAULT : ORB_PRIO_HIGH);
```
Comment (dagar) [inline src/drivers/imu/mpu6000/mpu6000.cpp]: Why is this necessary?

File: src/drivers/imu/amov_imu/amov_imu.cpp (hunk 3 above)
Comment (zarathustr) [reply src/drivers/imu/amov_imu/amov_imu.cpp]: OK, I got it

File: cmake/configs/nuttx_px4fmu-v2_default.cmake (hunk 4 above)
Comment (zarathustr) [reply cmake/configs/nuttx_px4fmu-v2_default.cmake]: Yes this is exactly tab from Qt 5.5.1. But I will check again later to prevent this

File: ROMFS/px4fmu_common/init.d/rc.sensors (hunk 6 above)
Comment (zarathustr) [reply ROMFS/px4fmu_common/init.d/rc.sensors]: This command is aimed to start the external IMU by default from the TELEM 2, which is the most common port for all autopilots.

File: src/drivers/imu/mpu6000/mpu6000.cpp (hunk 7 above)
Comment (zarathustr) [reply src/drivers/imu/mpu6000/mpu6000.cpp]: This is just an experimental revision for the sake of convenience that we can block the mpu6000 by priority swapping

**Comment by zarathustr** [comment]
@dagar Thanks dagar, I have tested the IMU on platforms including fmu-v2 fmu-v3 fmu-v4 fmu-v5, so switching to fmu-v3 is good i think

**Comment by zarathustr** [comment]
@dagar Hi Dagar! We have a good news that the AMOV IMU driver can be linked with default ld.script provided by the px4fmu-v2_default. So I think this time it is OK to include this merge for the fmu v2.

**Comment by huangwen0907** [comment]
@simonegu  @MaEtUgR  hi, do you have some ideas to solve the failure `.data' will not fit in regionflash'`  as  I have the same problems when I build in my OS? Thank you very much.

**Comment by zarathustr** [comment]
@huangwen0907 This is quite simple. Just compile px4fmu-v3_default and upload it to fmu v2. @dagar

**Comment by zarathustr** [comment]
@dagar The checks are all passed. Please review them and consider possible merge.

File: ROMFS/px4fmu_common/init.d/rc.sensors [hunk 8]
```diff
@@ -79,9 +79,25 @@ then
 
 	# Internal I2C bus
 	hmc5883 -C -T -I -R 4 start
-
+	
 	# Internal SPI bus ICM-20608-G
 	mpu6000 -T 20608 start
+	
+	# External UART DMA bus AMOV-IMU
+	if param compare SYS_COMPANION 1500000
```
Comment (dagar) [inline ROMFS/px4fmu_common/init.d/rc.sensors]: This should be in the main SYS_COMPANION section.

Comment (zarathustr) [reply ROMFS/px4fmu_common/init.d/rc.sensors]: But agar the configuration is not all the same for different autopilots

Comment (dagar) [reply ROMFS/px4fmu_common/init.d/rc.sensors]: It's starting a serial device connected to a special serial port that's handled consistently across boards. Why would this be any different?

Comment (zarathustr) [reply ROMFS/px4fmu_common/init.d/rc.sensors]: I mean there will be not only one external IMU so we really need more ports but they actually aren't consistent in various versions

**Comment by zarathustr** [comment]
@dagar I just don't know why the following codes are conflicting:

#define DRV_ACC_DEVTYPE_AMOV_IMU	0x63
#define DRV_GYR_DEVTYPE_AMOV_IMU	0x64
#define DRV_MAG_DEVTYPE_AMOV_IMU	0x65
#define DRV_BARO_DEVTYPE_AMOV_IMU	0x66

So I just temporarily move it into the local header instead of drv_sensor.h

**Comment by zarathustr** [comment]
@dagar The build seems to get wrong in build sitl and ros tests. But we indeed do not have configuration for the two systems. Can you figure out the reason?

File: ROMFS/px4fmu_common/init.d/rc.sensors [hunk 9]
```diff
@@ -79,7 +79,7 @@ then
 
 	# Internal I2C bus
 	hmc5883 -C -T -I -R 4 start
-
+
```
Comment (dagar) [inline ROMFS/px4fmu_common/init.d/rc.sensors]: whitespace

File: ROMFS/px4fmu_common/init.d/rc.sensors [hunk 10]
```diff
@@ -179,7 +179,7 @@ then
 fi
 
 if ver hwcmp PX4FMU_V4
-then
+then
```
Comment (dagar) [inline ROMFS/px4fmu_common/init.d/rc.sensors]: whitespace

**Comment by dagar** [comment]
Hi @zarathustr, I seem to have missed your last replies. If you can get this rebased on master and squashed we should be able to work through the remaining issues to get this merged. Check the whitespace differences via GitHub.

**Comment by TSC21** [comment]
@zarathustr can you rebase? Thanks!

**Comment by bys1123** [comment]
@zarathustr Are you still work on this?

**Comment by zarathustr** [comment]
@bys1123  Yes I am still working on it

**Comment by mhkabir** [comment]
@zarathustr is the hardware available somewhere?

**Comment by julianoes** [comment]
@zarathustr what's the status here? Is there hardware available so that it can be added to the CI hardware testing to make sure it's not broken once merged?

**Comment by zarathustr** [comment]
> @zarathustr what's the status here? Is there hardware available so that it can be added to the CI hardware testing to make sure it's not broken once merged?

Thanks for asking. We will revise for the new version and try to push to the master.

**Comment by zarathustr** [comment]
AMOV-IMU will no longer be supported. The AMOV Team has closed and will not participate such issue.

**Comment by zarathustr** [comment]
A new IMU branch will be opened in the near future that supports mainstream IMUs like Microstrain, Xsens and SBG

**Comment by aokholm** [comment]
@zarathustr What's the status on the new IMU branch? Do you have any more information regarding the proposed IMUS ect?
